src/Hosting/Hosting/src/Internal/HostingMetrics.cs

```diff
@@ -0,0 +1,42 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.Metrics;
+using Microsoft.Extensions.Metrics;
+
+namespace Microsoft.AspNetCore.Hosting;
+
+internal class HostingMetrics : IDisposable
+{
+    private readonly Meter _meter;
+    private readonly ObservableCounter<long> _totalRequestsCounter;
+    private readonly ObservableUpDownCounter<long> _currentRequestsCounter;
+    private readonly ObservableCounter<long> _failedRequestsCounter;
+
+    public HostingMetrics(IMetricsFactory metricsFactory)
+    {
+        _meter = metricsFactory.CreateMeter("Microsoft.AspNetCore.Hosting");
+
+        _totalRequestsCounter = _meter.CreateObservableCounter<long>(
+            "total-requests",
+            () => HostingEventSource.Log.TotalRequests,
+            description: "Total Requests");
+
+        _currentRequestsCounter = _meter.CreateObservableUpDownCounter<long>(
+            "current-requests",
+            () => HostingEventSource.Log.CurrentRequests,
+            description: "Current Requests");
+
+        _failedRequestsCounter = _meter.CreateObservableCounter<long>(
+            "failed-requests",
+            () => HostingEventSource.Log.FailedRequests,
+            description: "Failed Requests");
+    }
+
+    public void Dispose()
+    {
+        _meter.Dispose();
+    }
+
+    public bool IsEnabled() => _totalRequestsCounter.Enabled || _currentRequestsCounter.Enabled || _failedRequestsCounter.Enabled;
```

Comment:
Not a fan of this. Could split up into multiple properties and test them individually when calling various methods on `HostingEventSource`.

Alternatively, adding an event to counters that triggers whether a counter is listened to could be used to move this long test somewhere else, and `IsEnabled` just turns the calculated result.

Reply:
I agree this one looks a little weird. I think what happens here will depend first on whether we continue to use process-wide values in the EventSource singleton or we switch to tracking the counters per-HostingApplication.

Reply:
Counters are now per-host.

Reply:
This new factoring feels much easier to follow 👍